File: scripts/native/stage-driver.mjs
// Build-time only. Downloads a pinned official Rust binary, never Python/Node runtimes.
import { createHash } from 'node:crypto';
import { mkdtemp, mkdir, writeFile, readFile, chmod, rename, rm } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { execFileSync } from 'node:child_process';

const root = fileURLToPath(new URL('../../', import.meta.url));
const version = '0.20.0';
const platform = process.platform === 'darwin' ? 'darwin-universal.tar.gz'
  : process.platform === 'win32' ? `windows-${process.arch === 'arm64' ? 'arm64' : 'x86_64'}-binary.zip`
  : null;
if (!platform) throw new Error('Native desktop driver packaging supports macOS and Windows');
const archiveName = `cua-driver-rs-${version}-${platform}`;
const hashes = {
  'darwin-universal.tar.gz': 'd5e61fecebd9a620e50c2b8b608c8e7e8141f74c6faebc2ae9ef5d0d96cce7b8',
  'windows-x86_64-binary.zip': 'c020fefee01aacc174a27fea84a0cb77d47ef8290bfc772b3db7e3e06670d2b2',
  'windows-arm64-binary.zip': 'e8d47cb35c7a719f12c3012caa12f1166ecd1614548afe9b18a2c600116f8bde',
};
const destination = resolve(process.argv[2] ?? join(root, 'console/src-tauri/binaries/native-cua-driver'));
const temporary = await mkdtemp(join(tmpdir(), 'potato-native-driver-'));
try {
  const archive = join(temporary, archiveName);
  const url = `https://github.com/trycua/cua/releases/download/cua-driver-rs-v${version}/${archiveName}`;
  // curl is present on supported macOS/Windows build hosts. No shell interpolation.
  execFileSync(process.platform === 'win32' ? 'curl.exe' : 'curl', ['--fail', '--location', '--retry', '3', '--connect-timeout', '30', '--max-time', '300', '--output', archive, url], { stdio: 'inherit' });
  const digest = createHash('sha256').update(await readFile(archive)).digest('hex');
  if (digest !== hashes[platform]) throw new Error('Official driver archive checksum mismatch');
  const binaryName = process.platform === 'win32' ? 'cua-driver.exe' : 'cua-driver';
  const members = execFileSync('tar', ['-tf', archive], { encoding: 'utf8', maxBuffer: 2_000_000 }).split(/\r?\n/)
    .filter(name => name === binaryName || name.endsWith(`/${binaryName}`));
  // A binary extracted from an .app retains a signature bound to that bundle.
  // Select the standalone executable supplied in the same official archive.
  const standalone = members.filter(name => !name.includes('.app/Contents/MacOS/'));
  const member = standalone.length === 1 ? standalone[0] : null;
  if (!member || member.startsWith('-') || member.split('/').includes('..')) throw new Error('Ambiguous or invalid driver archive');
  const bytes = execFileSync('tar', ['-xOf', archive, member], { maxBuffer: 200_000_000 });
  const staged = join(temporary, binaryName);
  await writeFile(staged, bytes, { mode: 0o755 });
  if (process.platform === 'darwin') execFileSync('codesign', ['--verify', '--strict', staged], { stdio: 'inherit' });
  await mkdir(destination, { recursive: true });
  const pending = join(destination, `${binaryName}.pending`);
  await writeFile(pending, bytes, { mode: 0o755 });
  await chmod(pending, 0o755);
  await rename(pending, join(destination, binaryName));
  await writeFile(join(destination, 'VERSION'), `${version}\n`);
  console.log(`Verified native driver staged in ${destination}`);
} finally {
  await rm(temporary, { recursive: true, force: true });
}

File: native/potato-gpui/stage_driver.py
"""Stage the pinned official Rust driver; no application runtime dependencies."""
import argparse
import hashlib
from pathlib import Path
import shutil
import subprocess
import sys
import tarfile
import tempfile
import zipfile

VERSION = "0.24.0"
ARCHIVES = {
    "darwin": ("darwin-universal-binary.tar.gz", "31790cb49baa206f6455fbc259f8f83ae27e86be908f5c8cac5ec2f8521f8382"),
    "win32": ("windows-x86_64-binary.zip", "cc22d7a44ad526f779f2df7e6da053dd898ef8e5014b1ecfc01728645f691be0"),
}


def stage(destination, system=sys.platform, archive=None):
    suffix, digest = ARCHIVES[system]
    name = f"cua-driver-rs-{VERSION}-{suffix}"
    destination = Path(destination)
    with tempfile.TemporaryDirectory(prefix="potato-driver-") as temporary:
        source = Path(archive) if archive else Path(temporary) / name
        if archive is None:
            subprocess.run(["curl", "--fail", "--location", "--retry", "3",
                            "--connect-timeout", "30", "--max-time", "300", "--output", str(source),
                            f"https://github.com/trycua/cua/releases/download/cua-driver-rs-v{VERSION}/{name}"], check=True)
        if hashlib.sha256(source.read_bytes()).hexdigest() != digest:
            raise ValueError("Official computer driver archive checksum mismatch")
        binary = "cua-driver.exe" if system == "win32" else "cua-driver"
        # Extract only the standalone executable, never archive paths or SDK libraries.
        if system == "win32":
            with zipfile.ZipFile(source) as package:
                matches = [p for p in package.namelist() if p == binary or p.endswith("/" + binary)]
                if len(matches) != 1:
                    raise ValueError("Ambiguous driver archive")
                data = package.read(matches[0])
        else:
            with tarfile.open(source) as package:
                matches = [p for p in package.getmembers() if p.isfile() and p.name == binary]
                if len(matches) != 1:
                    raise ValueError("Missing standalone driver")
                data = package.extractfile(matches[0]).read()
        staged = Path(temporary) / binary
        staged.write_bytes(data)
        staged.chmod(0o755)
        if system == "darwin":
            # The pinned universal standalone archive has no usable enclosing
            # bundle signature. Sign the verified bytes as our nested helper.
            subprocess.run(["codesign", "--force", "--sign", "-", str(staged)], check=True)
            subprocess.run(["codesign", "--verify", "--strict", str(staged)], check=True)
        destination.mkdir(parents=True, exist_ok=True)
        shutil.copy2(staged, destination / binary)
        (destination / "VERSION").write_text(VERSION + "\n", encoding="utf-8")
    return destination / binary


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--dest", type=Path, default=Path(__file__).resolve().parent / "target/computer-driver")
    parser.add_argument("--archive", type=Path, help="Use a local archive, still checked against the pinned digest")
    args = parser.parse_args()
    print(stage(args.dest, archive=args.archive))
